src/sql/workbench/contrib/notebook/browser/calloutDialog/imageCalloutDialog.ts

```diff
@@ -0,0 +1,238 @@
+import { InputBox } from 'sql/base/browser/ui/inputBox/inputBox';
+import { Checkbox } from 'sql/base/browser/ui/checkbox/checkbox';
+import { RadioButton } from 'sql/base/browser/ui/radioButton/radioButton';
+import { DialogWidth } from 'sql/workbench/api/common/sqlExtHostTypes';
+import { attachModalDialogStyler } from 'sql/workbench/common/styler';
+
+
+export class ImageCalloutDialog extends CalloutDialog {
+	private _selectionComplete: Deferred<ICalloutDialogOptions>;
+	private _imageLocationLabel: HTMLElement;
+	private _imageLocalRadioButton: RadioButton;
+	private _editorImageLocationGroup: string = 'editorImageLocationGroup';
+	private _imageRemoteRadioButton: RadioButton;
+	private _imageUrlLabel: HTMLElement;
+	private _imageUrlInputBox: InputBox;
+	private _imageBrowseButton: HTMLAnchorElement;
+	private _imageEmbedLabel: HTMLElement;
+	private _imageEmbedCheckbox: Checkbox;
+	private readonly insertButtonText = localize('callout.insertButton', "Insert");
+	private readonly cancelButtonText = localize('callout.cancelButton', "Cancel");
+	private readonly locationLabel = localize('callout.locationLabel', "Image location");
+	private readonly localImageLabel = localize('callout.localImageLabel', "This computer");
+	private readonly remoteImageLabel = localize('callout.remoteImageLabel', "Online");
+	private readonly pathInputLabel = localize('callout.pathInputLabel', "Image URL");
+	private readonly pathPlaceholder = localize('callout.pathPlaceholder', "Enter image path");
+	private readonly urlPlaceholder = localize('callout.urlPlaceholder', "Enter image URL");
+	private readonly browseAltText = localize('callout.browseAltText', "Browse");
+	private readonly embedImageLabel = localize('callout.embedImageLabel', "Attach image to notebook");
+
+	constructor(
+		title: string,
+		width: DialogWidth,
+		dialogProperties: IDialogProperties,
+		@IPathService private readonly _pathService: IPathService,
+		@IFileDialogService private readonly _fileDialogService: IFileDialogService,
+		@IContextViewService private _contextViewService: IContextViewService,
+		@IThemeService themeService: IThemeService,
+		@ILayoutService layoutService: ILayoutService,
+		@IAdsTelemetryService telemetryService: IAdsTelemetryService,
+		@IContextKeyService contextKeyService: IContextKeyService,
+		@IClipboardService clipboardService: IClipboardService,
+		@ILogService logService: ILogService,
+		@ITextResourcePropertiesService textResourcePropertiesService: ITextResourcePropertiesService
+	) {
+		super(
+			title,
+			width,
+			dialogProperties,
+			themeService,
+			layoutService,
+			telemetryService,
+			contextKeyService,
+			clipboardService,
+			logService,
+			textResourcePropertiesService
+		);
+
+		this._selectionComplete = new Deferred<ICalloutDialogOptions>();
+	}
+
+	/**
+	 * Opens the dialog and returns a promise for what options the user chooses.
+	 */
+	public open(): Promise<ICalloutDialogOptions> {
```

Comment:
This seems like something that should still be a part of the base class - it's not doing any special logic here

Reply:
I moved this out to each custom calloutDialog type. What would it look like to exist more in more generic manner?

Reply:
You could use generics for that - have the base class definition be 

`export abstract class CalloutDialog<T> extends Modal {`

and then in that define the abstract function : 

`public abstract open(): Promise<T>`

The implementations would then look like this : 

`export class ImageCalloutDialog extends CalloutDialog<IImageCalloutDialogOptions> {`

and 

`public open(): Promise<ICalloutDialogOptions> {`